### PR TITLE
Kaito additional error & quota handling

### PR DESCRIPTION
**Pull request by tejhan**
- Added new & refactored existing error handling scenarios
- Implemented manual quota checking for more detailed user feedback
- Refactored helper functions & updated llama2 CRD templates
- Minor UX tweaks
- Refactored functions
- Added documentation

</img>

### CODE REVIEW

File: src/panels/KaitoModelsPanel.ts [hunk 1]
```diff
@@ -96,21 +101,103 @@ export class KaitoModelsPanelDataProvider implements PanelDataProvider<"kaitoMod
...
+                const getCommand = `get workspace workspace-${model}`;
+                getResult = await invokeKubectlCommand(this.kubectl, this.kubeConfigFilePath, getCommand);
+            });
+            if (getResult === null || failed(getResult)) {
+                await longRunning(`Verifying available subscription quota for deployment...`, async () => {
+                    const [gpuFamily, requiredCPUs] = this.parseGPU(gpu);
+                    void requiredCPUs;
+                    const credential = new DefaultAzureCredential();
+                    const computeClient = new ComputeManagementClient(credential, this.subscriptionId);
+                    const containerServiceClient = new ContainerServiceClient(credential, this.subscriptionId);
```
Comment (hsubramanianaks) [inline src/panels/KaitoModelsPanel.ts:151]: we have utility file to create aks client which can be used and refer: KaitoPanel.ts

```
this.containerServiceClient = getAksClient(sessionProvider, this.subscriptionId);
```

File: src/panels/KaitoModelsPanel.ts [hunk 2]
```diff
@@ -96,21 +101,103 @@ export class KaitoModelsPanelDataProvider implements PanelDataProvider<"kaitoMod
...
+            await longRunning(`Checking if workspace exists...`, async () => {
+                const getCommand = `get workspace workspace-${model}`;
+                getResult = await invokeKubectlCommand(this.kubectl, this.kubeConfigFilePath, getCommand);
+            });
+            if (getResult === null || failed(getResult)) {
+                await longRunning(`Verifying available subscription quota for deployment...`, async () => {
+                    const [gpuFamily, requiredCPUs] = this.parseGPU(gpu);
+                    void requiredCPUs;
+                    const credential = new DefaultAzureCredential();
+                    const computeClient = new ComputeManagementClient(credential, this.subscriptionId);
```
Comment (hsubramanianaks) [inline src/panels/KaitoModelsPanel.ts:150]: I would suggest a new utility method in file (src/commands/utils/arm.ts) to create ComputeManagementClient so that it can be used for future use cases as well. 

```
export function getComputeManagementClient(
    sessionProvider: ReadyAzureSessionProvider,
    subscriptionId: string,
): ComputeManagementClient {
    return new ComputeManagementClient(getCredential(sessionProvider), subscriptionId, { endpoint: getArmEndpoint() });
}
```

**Comment by Tatsinnit** [comment]
📓 Please Note: Thanks heaps for this, I think upon merger of this #1001 , Size of this PR will significantly decrease, so once #1001  is taken care off - I will review this sometime tomorrow. Thanks you.

File: src/panels/KaitoModelsPanel.ts [hunk 3]
```diff
@@ -96,21 +100,102 @@ export class KaitoModelsPanelDataProvider implements PanelDataProvider<"kaitoMod
     nullIsFalse(value: boolean | null): boolean {
         return value ?? false;
     }
+    parseGPU(gpuRequirement: string): [string, number] {
+        const regex = /^Standard_NC(\d+)(ads_A100_v4|s_v3)$/;
+        const match = gpuRequirement.match(regex);
+        if (match) {
+            const cpus = parseInt(match[1], 10);
+            let family: string;
+            switch (match[2]) {
```
Comment (Tatsinnit) [inline src/panels/KaitoModelsPanel.ts:109]: 💡 Observation and highly recommended: For our own readability sake and for future better understanding let's comment this specific part and mention a sample value (fake sample value) for `match[2]` ?

Relevance with in code level comment, please. = )

File: src/panels/KaitoModelsPanel.ts [hunk 4]
```diff
@@ -96,21 +100,102 @@ export class KaitoModelsPanelDataProvider implements PanelDataProvider<"kaitoMod
...
+                    family = "standardNCSv3Family";
+                    break;
+                default:
+                    throw new Error("Unknown gpu format");
+            }
+
+            return [family, cpus];
+        } else {
+            throw new Error("Unknown gpu format");
+        }
```
Comment (Tatsinnit) [inline src/panels/KaitoModelsPanel.ts:123]: 💡 minor suggestions: either do it in separate PR for update but line-116 and line 122 both have same return set, which could be either extracted and return again, just a note only sake. (In case you notice more deeper pattern)

File: src/panels/KaitoModelsPanel.ts [hunk 5]
```diff
@@ -96,21 +100,102 @@ export class KaitoModelsPanelDataProvider implements PanelDataProvider<"kaitoMod
...
     }
+    parseGPU(gpuRequirement: string): [string, number] {
+        const regex = /^Standard_NC(\d+)(ads_A100_v4|s_v3)$/;
+        const match = gpuRequirement.match(regex);
+        if (match) {
+            const cpus = parseInt(match[1], 10);
+            let family: string;
+            switch (match[2]) {
+                case "ads_A100_v4":
+                    family = "StandardNCADSA100v4Family";
```
Comment (Tatsinnit) [inline src/panels/KaitoModelsPanel.ts:111]: 💡 if this is too many times use, we can `enum` `case` types like `ads_A100_v4` et. al. for better handling.

File: src/panels/KaitoModelsPanel.ts (hunk 3 above)
Comment (Tatsinnit) [inline src/panels/KaitoModelsPanel.ts:109]: 💡 Note: something of this sort for `enum`

```*.ts
enum GpuFamily {
    ads_A100_v4 = "StandardNCADSA100v4Family",
    s_v3 = "standardNCSv3Family"
}
```

File: src/panels/KaitoModelsPanel.ts [hunk 6]
```diff
@@ -96,21 +100,102 @@ export class KaitoModelsPanelDataProvider implements PanelDataProvider<"kaitoMod
...
+        yaml: string,
+        gpu: string,
+        webview: MessageSink<ToWebViewMsgDef>,
+    ) {
         this.cancelToken = false;
+        if (this.checkingGPU) {
+            return;
+        }
         try {
+            this.checkingGPU = true;
```
Comment (Tatsinnit) [inline src/panels/KaitoModelsPanel.ts:137]: 💡 Highly recommended: Can you please comment this function with the intent and the usage of various things like setting `this.checkingGPU = true` 

something like:

\** 

Intent of this function is to handleDeploy with few default values set ...... <Explanation for latter understanding>

**/

File: src/panels/KaitoModelsPanel.ts [hunk 7]
```diff
@@ -96,21 +100,102 @@ export class KaitoModelsPanelDataProvider implements PanelDataProvider<"kaitoMod
...
+                    );
+                    const quotas = computeClient.usageOperations.list(cluster.location);
+                    let foundQuota = null;
+                    for await (const quota of quotas) {
+                        if (quota.name.value === gpuFamily) {
+                            foundQuota = quota;
+                            break;
+                        }
+                    }
+                    if (!foundQuota || !(foundQuota.currentValue + requiredCPUs <= foundQuota.limit)) {
```
Comment (Tatsinnit) [inline src/panels/KaitoModelsPanel.ts:162]: 💡 lets extract this from one big function to something like this? What do you think?

- Helper Functions: `findMatchingQuota`, `isQuotaSufficient`, and `promptForQuotaIncrease` have been extracted for clarity.
- Error Handling: The `promptForQuotaIncrease` function encapsulates the error message display and handling of user interaction.

```*.ts
if (getResult === null || failed(getResult)) {
    await longRunning(`Verifying available subscription quota for deployment...`, async () => {
        const [gpuFamily, requiredCPUs] = this.parseGPU(gpu);
        const computeClient = getComputeManagementClient(this.sessionProvider, this.subscriptionId);
        const containerServiceClient = getAksClient(this.sessionProvider, this.subscriptionId);
        const cluster = await containerServiceClient.managedClusters.get(
            this.resourceGroupName,
            this.clusterName,
        );
        const quotas = computeClient.usageOperations.list(cluster.location);
        
        const foundQuota = await findMatchingQuota(quotas, gpuFamily);
        
        if (!foundQuota || !isQuotaSufficient(foundQuota, requiredCPUs)) {
            await promptForQuotaIncrease();
        } else {
            quotaAvailable = true;
        }
    });
} else {
    quotaAvailable = true;
}

// Helper functions
async function findMatchingQuota(quotas, gpuFamily) {
    for await (const quota of quotas) {
        if (quota.name.value === gpuFamily) {
            return quota;
        }
    }
    return null;
}

function isQuotaSufficient(quota, requiredCPUs) {
    return quota.currentValue + requiredCPUs <= quota.limit;
}

async function promptForQuotaIncrease() {
    const selection = await vscode.window.showErrorMessage(
        `Your current Azure subscription doesn't have enough quota to deploy this model. Proceed to request a quota increase.`,
        "Learn More"
    );

    if (selection === "Learn More") {
        vscode.env.openExternal(
            vscode.Uri.parse("https://learn.microsoft.com/en-us/azure/quotas/quickstart-increase-quota-portal")
        );
    }
}

```

File: webview-ui/src/Kaito/KaitoModels.tsx [hunk 8]
```diff
@@ -89,13 +89,20 @@ export function KaitoModels(initialState: InitialState) {
...
-    function generateYAML(model: string) {
+    function generateYAML(model: string): [string, string | undefined] {
         const modelDetails = getModelDetails(model);
         const name = stringOrUndefined(modelDetails?.modelName);
         const gpu = stringOrUndefined(modelDetails?.minimumGpu);
         let appname = name;
         let metadataname = name;
         let inferencename = name;
+        let privateConfig = "";
+        if (name.substring(0, 7) === "llama-2") {
```
Comment (Tatsinnit) [inline webview-ui/src/Kaito/KaitoModels.tsx:100]: 💡 Let's definitely code comment things like this for future understanding, please. :)

File: webview-ui/src/Kaito/KaitoModels.tsx [hunk 9]
```diff
@@ -115,19 +122,21 @@ resource:
...
   preset:
-    name: ${inferencename}`;
-        return yaml;
+    name: ${inferencename}${privateConfig}`;
+        return [yaml, gpu];
     }
 
     function generateCRD(model: string) {
-        const yaml = generateYAML(model);
+        const yaml = generateYAML(model)[0];
```
Comment (Tatsinnit) [inline webview-ui/src/Kaito/KaitoModels.tsx:130]: 💡 Quick note: What is the relevance of `generateYAML(model)[0]` what standard value `[0]` will have and add sample value as comment for cahnces if this contract gets violated in future at least we would know what was the extended valuer.

File: webview-ui/src/Kaito/KaitoModels.tsx [hunk 10]
```diff
@@ -115,19 +122,21 @@ resource:
...
         vscode2.postGenerateCRDRequest({ model: yaml });
         return;
     }
 
     function onClickDeployKaito(model: string) {
-        const yaml = generateYAML(model);
-        vscode2.postDeployKaitoRequest({ model: model, yaml: yaml });
+        const [yaml, gpu] = generateYAML(model);
+        if (!(gpu === undefined)) {
+            vscode2.postDeployKaitoRequest({ model: model, yaml: yaml, gpu: gpu });
```
Comment (Tatsinnit) [inline webview-ui/src/Kaito/KaitoModels.tsx:138]: 💡 Can we please avoid `numeric` like `vscode2` value if that is not the intent for contextual understanding? (We can come out with more understandable and human readable variable names here :) 

Let definitely refactor this sooner please. 

things like this should have better names:

```*.ts
export const vscode2 = getWebviewMessageContext<"kaitoModels">({
    generateCRDRequest: null,
    deployKaitoRequest: null,
    workspaceExistsRequest: null,
    updateStateRequest: null,
    resetStateRequest: null,
    cancelRequest: null,
});
```

**Review by Tatsinnit — APPROVED** [review]
❤️ Thank you so much for this, looks good, there are few recommendation and if those are easy for you to make them quickly without too much turn-around time then please do so, otherwise we can handle in the follow up PR. 

- Lets also most definitely comment on some part of code which need relevance for better code readability for now.

- Lets get this tested as well please, do share away latest `*.vsix` and for that might be gentle fyi to @qpetraroia 

Rest looks good to me.

**Comment by tejhan** [comment]
Vsix for testing ->  [vscode-aks-tools-1.4.11-kaito-v6.zip](https://github.com/user-attachments/files/17467858/vscode-aks-tools-1.4.11-kaito-v6.zip)

**Review by Tatsinnit — APPROVED** [review]

File: src/commands/aksKaito/aksKaitoCreateCRD.ts [hunk 11]
```diff
@@ -12,8 +12,22 @@ import { getExtension, longRunning } from "../utils/host";
 
 export default async function aksKaitoCreateCRD(_context: IActionContext, target: unknown): Promise<void> {
     const cloudExplorer = await k8s.extension.cloudExplorer.v1;
+    if (!cloudExplorer.available) {
```
Comment (hsubramanianaks) [inline src/commands/aksKaito/aksKaitoCreateCRD.ts:15]: cloud explorer check is repeated, can you remove one?

File: src/panels/KaitoModelsPanel.ts [hunk 12]
```diff
@@ -166,35 +292,12 @@ export class KaitoModelsPanelDataProvider implements PanelDataProvider<"kaitoMod
...
-                return {
-                    clusterName: this.clusterName,
-                    modelName: "",
-                    workspaceExists: false,
-                    resourceReady: null,
-                    inferenceReady: null,
-                    workspaceReady: null,
-                    age: 0,
-                } as InitialState;
+                return this.getInitialState();
```
Comment (hsubramanianaks) [inline src/panels/KaitoModelsPanel.ts:295]: Suggestion - This can be done as separate PR @tejhan where you can add telemetry events to track failures/success which will help drive more enhancements/bug fixes for this feature.

**Review by hsubramanianaks — DISMISSED** [review]
@tejhan looks good to me, just couple of suggestions and regarding telemetry events we can add few just to track failures especially so that we can be proactive on finding user errors.

File: src/commands/aksKaito/aksKaitoCreateCRD.ts (hunk 11 above)
Comment (tejhan) [reply src/commands/aksKaito/aksKaitoCreateCRD.ts:15]: The second one was deleted 👍, moved things around .

File: src/panels/KaitoModelsPanel.ts (hunk 12 above)
Comment (tejhan) [reply src/panels/KaitoModelsPanel.ts:295]: Makes sense, will look into including this for next PR

**Review by hsubramanianaks — APPROVED** [review]
LGTM , tested with Tats and Reinier on dev cluster (Automatic cluster doesn't work , we can create an issue and move on). Thank you @tejhan for great work.

**Review by Tatsinnit — APPROVED** [review]
❤️ Thank you so much for this, and adding documentation. Let's get this in. We are aware about a tiny glitch with automatic cluster but that's known. We can handle this latter.  I have tested this in my local and looks good. Thanks.

Gentle fyi to @hsubramanianaks, @ReinierCC and @qpetraroia + @tejhan 

Thank you so much guys!!
